lcls_tools/common/data_analysis/bmad_modeling.py

```diff
@@ -0,0 +1,573 @@
+            "ele.a.eta",
+            "ele.a.etap",
+            "ele.a.gamma",
+            "ele.a.phi",
+            "ele.b.beta",
+            "ele.b.alpha",
+            "ele.b.eta",
+            "ele.b.etap",
+            "ele.b.gamma",
+            "ele.b.phi",
+            "ele.x.eta",
+            "ele.x.etap",
+            "ele.y.eta",
+            "ele.y.etap",
+            "ele.s",
+            "ele.l",
+            "ele.e_tot",
+            "ele.p0c",
+            "ele.mat6",
+            "ele.vec0",
+        ]
+        output = {k: tao.lat_list("*", k) for k in outkeys}
+        return output
+
+    def get_element(self, tao, datum):
+        return tao.data_parameter(datum, "ele_name")[0].split(";")[1]
+
+    def update_datum_meas(self, tao, datum, useDesing=True):
+        """Updates datum with EPICS or design twiss values"""
+        element = self.get_element(tao, datum)
+        TWISS = ["beta_a", "alpha_a", "beta_b", "alpha_b"]
+        cmd = []
+        if useDesing:
+            ele_twiss = tao.ele_twiss(element, which="design")
+            for ii, p in enumerate(TWISS):
+                cmd.append(f"set dat {datum}[{ii+1}]|meas = {ele_twiss[p]}")
+        else:
+            device = tao.ele_head(element)["alias"]
+            EPICS_ATTRIBUTE = ["BETA_X", "ALPHA_X", "BETA_Y", "ALPHA_Y"]
+            pvs = [device + ":" + twiss for twiss in EPICS_ATTRIBUTE]
+            if self.data_source in ["ACT", "DES"]:
+                measured = self.get_live(pvs)
+            elif self.data_source == "ARCHIVE":
+                measured = lcls_archiver_restore(pvs, self.date_time)
+            for ii, pv in enumerate(pvs):
+                cmd.append(
+                    f"set dat {datum}[{ii+1}]|meas = {measured[pvs[ii]]}")
+        tao.cmds(cmd)
+        self.tc(tao, f"show dat {datum}")
+
+    def match_twiss(self, tao, variable, datum):
+        match_cmds = ["veto var *", "veto dat *@*",
+                      "set global n_opti_cycles = 912"]
+        match_cmds.append(f"use var {variable}[1:4]")
+        match_cmds.append(f"use dat {datum}[1:4]")
+        tao.cmds(match_cmds)
+        tao.cmd("run")
+        element = self.get_element(tao, datum)
+        self.show_twiss(tao, element, [])
+
+    def get_twiss(self, tao, element, which="design"):
+        """Gets twiss for element, which can be model, desing or base"""
+        result = tao.ele_twiss(element, which=which)
+        return [result[p] for p in ["beta_a", "alpha_a", "beta_b", "alpha_b"]]
+
+    def bmag(self, twiss, twiss_reference):
+        """Calculates BMAG from imput twiss and reference twiss"""
+        beta_a, alpha_a, beta_b, alpha_b = twiss
+        beta_a_ref, alpha_a_ref, beta_b_ref, alpha_b_ref = twiss_reference
+        bmag_a = self.bmag_func(beta_a, alpha_a, beta_a_ref, alpha_a_ref)
+        bmag_b = self.bmag_func(beta_b, alpha_b, beta_b_ref, alpha_b_ref)
+        return (bmag_a, bmag_b)
+
+    def bmag_func(self, bb, ab, bl, al):
```

Comment:
this is a more general function, can prob be pulled into another file for others to use as well. Can you add a comment here on inputs?

Reply:
Done major rework.  Let's take a look at new pull request and see.